stan/math/prim/scal/prob/normal_lcdf.hpp

```diff
@@ -53,34 +55,132 @@ inline return_type_t<T_y, T_loc, T_scale> normal_lcdf(const T_y& y,
     const T_partials_return scaled_diff
         = (y_dbl - mu_dbl) / (sigma_dbl * SQRT_2);
 
-    T_partials_return one_p_erf;
-    if (scaled_diff < -37.5 * INV_SQRT_2) {
-      one_p_erf = 0.0;
-    } else if (scaled_diff < -5.0 * INV_SQRT_2) {
-      one_p_erf = erfc(-scaled_diff);
-    } else if (scaled_diff > 8.25 * INV_SQRT_2) {
-      one_p_erf = 2.0;
+    const T_partials_return sigma_sqrt2 = sigma_dbl * SQRT_2;
+    const T_partials_return x2 = square(scaled_diff);
+
+    // check whether scaled_diff^10 term will overflow
+    const bool overflow_check
+        = 10.0 * log(-scaled_diff)
+          < log(std::numeric_limits<T_partials_return>::max());
+
+    // use erfc() instead of erf() in order to retain precision
+    // since for x>0 erfc()->0
+    if (scaled_diff > 0.0) {
```

Comment:
It looks like this function is piecewise defined over four intervals:

```
(-inf, o), (o, -20), (-20, 0), (0, inf)
```

Where ```o = pow(std::numeric_limits<T_partials_return>::max(), 0.1)```

The overflow_check logic was a bit confusing. Can this be rearranged to just be 4 flat if-else statements? Instead of three if-elses with an overflow check condition in the third one?

Reply:
Yes that makes sense. We can simply replace

`} else {`
`      if (overflow_check) {`

with 

`} else if (10.0 * log(-scaled_diff)`
          `< log(std::numeric_limits<T_partials_return>::max())) {`